src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs

```diff
@@ -1308,6 +1313,7 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
                     if (addValue)
                     {
                         tagBuilder.MergeAttribute("value", attributeValue, replaceExisting: isExplicitValue);
+                        addMaxLengthAttribute = true;
```

Comment:
As-is, will add a `maxLength` attribute if the `IHtmlGenerator` call was `GenerateHidden(...)` or if `htmlAttributes` contained e.g. `type="color"`. The `maxLenth` attribute applies only when the `type` is `email`, `password`, `search`, `telephone`, `text`, or `url`. And, `typeAttributeString` doesn't cover the case where `htmlAttributes` is `null` (and the `type` comes from the caller).

Suggest adding a separate conditional block or perhaps a `switch` using `suppliedTypeString` and the `type`s where `maxLength` applies.

----
As a minor detail, `typeAttributeValue` and `typeAttributeString` shouldn't be necessary. They're specific to the case where the user supplied attributes but `suppliedTypeString` already reflects `htmlAttributes`.